File: frontend/scripts/content-coverage.py
#!/usr/bin/env python3
"""Content Coverage Audit - compares syllabus topics against actual content files."""
import os
import re

def parse_syllabus():
    """Parse lib/syllabus.ts to extract all topics."""
    with open('lib/syllabus.ts', 'r', encoding='utf-8') as f:
        content = f.read()

    lines = content.split('\n')
    entries = []

    current_class_slug = ''
    current_subject_slug = ''
    current_unit_slug = ''
    current_unit_title = ''
    in_topics = False
    topic_buffer = []

    i = 0
    while i < len(lines):
        line = lines[i]

        # Class level (4 spaces) - slug on one line, name on next
        if re.match(r'^    slug: "([^"]+)"', line):
            current_class_slug = re.match(r'^    slug: "([^"]+)"', line).group(1)
            # Look for name on next line
            if i + 1 < len(lines) and 'name:' in lines[i + 1]:
                pass
            i += 1
            continue

        # Subject level (8 spaces) - slug on one line, name on next
        if re.match(r'^        slug: "([^"]+)"', line):
            current_subject_slug = re.match(r'^        slug: "([^"]+)"', line).group(1)
            if i + 1 < len(lines) and 'name:' in lines[i + 1]:
                pass
            i += 1
            continue

        # Unit level (12 spaces)
        unit_match = re.match(r'^            id: "([^"]+)"', line)
        if unit_match:
            current_unit_slug = unit_match.group(1)
            # Look for title in next few lines
            for j in range(i + 1, min(i + 5, len(lines))):
                title_match = re.search(r'title: "([^"]+)"', lines[j])
                if title_match:
                    current_unit_title = title_match.group(1)
                    break
            i += 1
            continue

        # Topics array
        if 'topics: [' in line and current_unit_slug:
            in_topics = True
            topic_buffer = []
            i += 1
            continue

        # Collect topics (14 spaces)
        if in_topics:
            topic_match = re.match(r'^              "([^"]+)"', line)
            if topic_match:
                topic_buffer.append(topic_match.group(1))
            elif line.strip() in ['],', ']']:
                in_topics = False
                if topic_buffer and current_subject_slug:
                    entries.append({
                        'class_slug': current_class_slug,
                        'subject_slug': current_subject_slug,
                        'unit_slug': current_unit_slug,
                        'unit_title': current_unit_title,
                        'topics': topic_buffer.copy(),
                    })
                current_unit_slug = ''
                current_unit_title = ''
                topic_buffer = []
            i += 1
            continue

        i += 1

    return entries

def get_content_files(base_path):
    """Get all content files with their paths."""
    files = []

    def scan_dir(directory, rel_path=''):
        try:
            entries = os.listdir(directory)
            for entry in entries:
                full_path = os.path.join(directory, entry)
                new_rel = os.path.join(rel_path, entry) if rel_path else entry
                if os.path.isdir(full_path):
                    scan_dir(full_path, new_rel)
                elif entry.endswith('.json') and 'mindmap' not in entry:
                    slug = re.sub(r'^\d+-', '', entry).replace('.json', '')
                    parts = new_rel.split(os.sep)
                    files.append({
                        'subject': parts[0],
                        'unit': parts[1] if len(parts) > 1 else '',
                        'slug': slug,
                        'path': full_path,
                    })
        except Exception as e:
            pass

    scan_dir(base_path)
    return files

def normalize(s):
    """Normalize string for comparison."""
    return re.sub(r'[^a-z0-9]', '-', s.lower()).replace('-+', '-').strip('-')

def is_covered(topic, content_files):
    """Check if a topic is covered by any content file."""
    t = normalize(topic)
    for f in content_files:
        c = normalize(f['slug'])
        if t == c or c in t or t in c:
            return True
        if len(t) > 10 and t[:20] in c:
            return True
    return False

# Main
print("=" * 60)
print("CONTENT COVERAGE AUDIT")
print("=" * 60)

syllabus = parse_syllabus()
total_topics = sum(len(e['topics']) for e in syllabus)
print(f"\nSyllabus entries: {len(syllabus)}")
print(f"Total topics: {total_topics}")

# Get content files
content_path = 'public/data/syllabus-notes'
content_files = get_content_files(content_path)
print(f"Content files found: {len(content_files)}")

# Group by subject
subjects = {}
for entry in syllabus:
    subj = entry['subject_slug']
    if subj not in subjects:
        subjects[subj] = []
    subjects[subj].append(entry)

# Get content by subject
subject_files = {}
for f in content_files:
    subj = f['subject']
    if subj not in subject_files:
        subject_files[subj] = []
    subject_files[subj].append(f)

# Generate report
grand_total = 0
grand_covered = 0

print("\n" + "=" * 60)
print("SUBJECT BREAKDOWN")
print("=" * 60)

for subject in ['biology', 'chemistry', 'english', 'mathematics', 'nepali', 'physics']:
    entries = subjects.get(subject, [])
    files = subject_files.get(subject, [])

    total_topic_count = sum(len(e['topics']) for e in entries)
    covered_count = 0
    gaps = []

    for entry in entries:
        for topic in entry['topics']:
            if is_covered(topic, files):
                covered_count += 1
            else:
                gaps.append(f"{entry['unit_slug']}: {topic[:50]}...")

    coverage = (covered_count / total_topic_count * 100) if total_topic_count > 0 else 0
    grand_total += total_topic_count
    grand_covered += covered_count

    print(f"\n[{subject.upper()}]")
    print(f"  Content files: {len(files)}")
    print(f"  Coverage: {covered_count}/{total_topic_count} ({coverage:.1f}%)")
    if gaps:
        print(f"  Missing: {len(gaps)} topics")
        print(f"  Sample gaps:")
        for g in gaps[:3]:
            print(f"    - {g}")

print("\n" + "=" * 60)
print("OVERALL SUMMARY")
print("=" * 60)
total_coverage = (grand_covered / grand_total * 100) if grand_total > 0 else 0
print(f"TOTAL: {grand_covered}/{grand_total} topics covered ({total_coverage:.1f}%)")
print(f"Missing: {grand_total - grand_covered} topics")
print("=" * 60)

File: frontend/scripts/content-coverage.ts
/**
 * Content Coverage Audit
 * Compares syllabus topics against actual content files to find gaps.
 */
import { readFileSync, readdirSync, statSync } from 'fs';
import { join } from 'path';

// Parse syllabus.ts to extract all topics
function parseSyllabus() {
  const content = readFileSync('lib/syllabus.ts', 'utf8').replace(/\r\n/g, '\n');
  const lines = content.split('\n');

  const result: Array<{
    classSlug: string;
    classTitle: string;
    subjectSlug: string;
    subjectTitle: string;
    unitSlug: string;
    unitTitle: string;
    topics: string[];
  }> = [];

  let currentClassSlug = '';
  let currentClassTitle = '';
  let currentSubjectSlug = '';
  let currentSubjectTitle = '';
  let currentUnitSlug = '';
  let currentUnitTitle = '';
  let inTopics = false;
  let topicBuffer: string[] = [];

  for (let i = 0; i < lines.length; i++) {
    const line = lines[i];

    // Detect class slug/name (4 spaces)
    const clsMatch = line.match(/^\s{4}slug:\s*"([^"]+)"\s*,\s*name:\s*"([^"]+)"/);
    if (clsMatch && !currentClassSlug) {
      currentClassSlug = clsMatch[1];
      currentClassTitle = clsMatch[2];
      console.log('Found class:', currentClassSlug);
      continue;
    }

    // Detect subject slug/name (8 spaces)
    const subjMatch = line.match(/^\s{8}slug:\s*"([^"]+)"\s*,\s*name:\s*"([^"]+)"/);
    if (subjMatch) {
      currentSubjectSlug = subjMatch[1];
      currentSubjectTitle = subjMatch[2];
      console.log('Found subject:', currentSubjectSlug);
      continue;
    }

    // Detect unit id (12 spaces)
    const unitIdMatch = line.match(/^\s{12}id:\s*"([^"]+)"/);
    if (unitIdMatch) {
      currentUnitSlug = unitIdMatch[1];
      // Look for title in next few lines
      for (let j = i + 1; j < Math.min(i + 5, lines.length); j++) {
        const titleMatch = lines[j].match(/\s*title:\s*"([^"]+)"/);
        if (titleMatch) {
          currentUnitTitle = titleMatch[1];
          break;
        }
      }
      continue;
    }

    // Detect topics array start
    if (line.includes('topics: [') && currentUnitSlug) {
      inTopics = true;
      topicBuffer = [];
      continue;
    }

    // Collect topic strings (14 spaces)
    if (inTopics) {
      const topicMatch = line.match(/^\s{14}"(.+)"(?:,|$)/);
      if (topicMatch) {
        topicBuffer.push(topicMatch[1]);
      } else if (line.trim() === '],') {
        inTopics = false;
        if (topicBuffer.length > 0 && currentSubjectSlug) {
          result.push({
            classSlug: currentClassSlug,
            classTitle: currentClassTitle,
            subjectSlug: currentSubjectSlug,
            subjectTitle: currentSubjectTitle,
            unitSlug: currentUnitSlug,
            unitTitle: currentUnitTitle,
            topics: topicBuffer,
          });
        }
        currentUnitSlug = '';
        currentUnitTitle = '';
        topicBuffer = [];
      }
    }
  }

  return result;
}

// Get all content file slugs from a directory
function getContentSlugs(basePath: string): Map<string, number> {
  const slugMap = new Map<string, number>();

  function scanDir(dir: string) {
    try {
      const entries = readdirSync(dir, { withFileTypes: true });
      for (const entry of entries) {
        const fullPath = join(dir, entry.name);
        if (entry.isDirectory()) {
          scanDir(fullPath);
        } else if (entry.name.endsWith('.json') && !entry.name.includes('mindmap')) {
          const slug = entry.name.replace(/^\d+-/, '').replace('.json', '');
          slugMap.set(slug, (slugMap.get(slug) || 0) + 1);
        }
      }
    } catch (e) {
      // ignore
    }
  }

  scanDir(basePath);
  return slugMap;
}

// Normalize slug for comparison
function normalize(s: string): string {
  return s.toLowerCase().replace(/[^a-z0-9]/g, '-').replace(/-+/g, '-').replace(/^-|-$/g, '');
}

// Check if a topic is covered by any content file
function isTopicCovered(topic: string, contentSlugs: Map<string, number>): boolean {
  const t = normalize(topic);
  for (const slug of contentSlugs.keys()) {
    const c = normalize(slug);
    if (t === c) return true;
    if (c.includes(t) || t.includes(c)) return true;
    // Check partial match (first 20 chars)
    if (t.length > 10 && c.includes(t.substring(0, 20))) return true;
    if (t.length > 10 && t.includes(c.substring(0, 20))) return true;
  }
  return false;
}

// Main
const topics = parseSyllabus();
console.log(`\nTotal syllabus entries parsed: ${topics.length}`);

// Group by subject
const subjectData: Record<string, typeof topics> = {};
for (const t of topics) {
  if (!subjectData[t.subjectSlug]) subjectData[t.subjectSlug] = [];
  subjectData[t.subjectSlug].push(t);
}

// Check each subject
const results: Array<{
  subject: string;
  subjectTitle: string;
  totalUnits: number;
  totalTopics: number;
  coveredTopics: number;
  gapCount: number;
  gaps: string[];
}> = [];

for (const [subject, entries] of Object.entries(subjectData)) {
  const contentPath = join('public/data/syllabus-notes', subject);
  let contentSlugs = new Map<string, number>();
  try {
    if (statSync(contentPath).isDirectory()) {
      contentSlugs = getContentSlugs(contentPath);
    }
  } catch {
    // directory doesn't exist
  }

  let totalTopics = 0;
  let coveredTopics = 0;
  const gaps: string[] = [];

  for (const entry of entries) {
    totalTopics += entry.topics.length;
    for (const topic of entry.topics) {
      if (isTopicCovered(topic, contentSlugs)) {
        coveredTopics++;
      } else {
        gaps.push(`${entry.unitSlug}/${topic.substring(0, 40)}...`);
      }
    }
  }

  const coverage = totalTopics > 0 ? (coveredTopics / totalTopics * 100) : 0;
  console.log(`\n=== ${subject} (${contentSlugs.size} content files) ===`);
  console.log(`  Units: ${entries.length}, Topics: ${totalTopics}, Covered: ${coveredTopics}, Gaps: ${totalTopics - coveredTopics} (${coverage.toFixed(1)}%)`);
  if (gaps.length > 0) {
    console.log(`  Sample gaps: ${gaps.slice(0, 5).join(', ')}${gaps.length > 5 ? ` ...(+${gaps.length - 5})` : ''}`);
  }
  results.push({
    subject,
    subjectTitle: entries[0]?.subjectTitle || '',
    totalUnits: entries.length,
    totalTopics,
    coveredTopics,
    gapCount: totalTopics - coveredTopics,
    gaps,
  });
}

console.log('\n=== OVERALL SUMMARY ===');
let grandTotal = 0, grandCovered = 0;
for (const r of results) {
  grandTotal += r.totalTopics;
  grandCovered += r.coveredTopics;
  console.log(`${r.subject}: ${r.coveredTopics}/${r.totalTopics} (${((r.coveredTopics / r.totalTopics) * 100).toFixed(1)}%)`);
}
console.log(`TOTAL: ${grandCovered}/${grandTotal} (${((grandCovered / grandTotal) * 100).toFixed(1)}%)`);
